drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+
+	virtio_pci_dev_reset(priv->vpdev);
+
+	/* disable all queue*/
+	for (i = 0; i < priv->nr_virtqs; i++) {
+		if (priv->vrings[i]->enable)
+			virtio_vdpa_vring_state_set(vid, i, 0);
+	}
+
+	/* Tell the host we've noticed this device. */
+	virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_ACK);
+
+	/* Tell the host we've known how to drive the device. */
+	virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_DRIVER);
+
+	priv->configured = 0;
+
+	/* The mutex may stay locked after event thread cancel - initiate it. */
+	pthread_mutex_init(&priv->vq_config_lock, NULL);
+	DRV_LOG(INFO, "vDPA device %d was closed.", vid);
+	return ret;
+}
+
+static int
+virtio_vdpa_dev_config(int vid)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	int ret, fd;
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	if (priv->configured) {
+		DRV_LOG(ERR, "Failed to reconfigure vid %d.", vid);
+		return -EINVAL;
+	}
+
+	priv->nr_virtqs = rte_vhost_get_vring_num(vid);
+
+	priv->vid = vid;
+	ret = virtio_vdpa_dma_map(priv);
+	if(ret) {
+		DRV_LOG(ERR, "Fail to do dma map: %d",ret);
+		return ret;
+	}
+	priv->configured = 1;
+	DRV_LOG(INFO, "vDPA device %d was configured.", vid);
+
+	fd =rte_intr_fd_get(priv->pdev->intr_handle);
+
+	if (priv->nvec <= (priv->nr_virtqs +1)) {
+		DRV_LOG(ERR, "Error dev interrupts %d less than queue: %d",
+			priv->nvec, priv->nr_virtqs +1);
+		return -EINVAL;
+	}
+
+	ret = virtio_pci_dev_interrupts_enable(priv->vpdev, fd, 0);
+	if (ret) {
+		DRV_LOG(ERR, "Error enabling virtio dev interrupts: %s",
+				strerror(errno));
+		return ret;
+	}
+
+	virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_FEATURES_OK);
+	/* Start the device */
+	virtio_pci_dev_set_status(priv->vpdev, VIRTIO_CONFIG_STATUS_DRIVER_OK);
+	DRV_LOG(INFO, "vDPA device %d  move to driver ok", vid);
+
+	return 0;
+}
+
+static int
+virtio_vdpa_group_fd_get(int vid)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	return priv->vfio_group_fd;
+}
+
+static int
+virtio_vdpa_device_fd_get(int vid)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+
+	if (priv == NULL) {
+		DRV_LOG(ERR, "Invalid vDPA device: %s", vdev->device->name);
+		return -ENODEV;
+	}
+	return priv->vfio_dev_fd;
+}
+
+static int
+virtio_vdpa_notify_area_get(int vid, int qid, uint64_t *offset, uint64_t *size)
+{
+	struct rte_vdpa_device *vdev = rte_vhost_get_vdpa_device(vid);
+	struct virtio_vdpa_priv *priv =
+		virtio_vdpa_find_priv_resource_by_vdev(vdev);
+	struct vfio_region_info reg = { .argsz = sizeof(reg) };
```

Comment:
this functionality of vfio stays in the lower layer.

Please move and provide an API.

Reply:
vfio_region_info removed